invokeai/app/services/shared/invocation_context.py

```diff
@@ -0,0 +1,625 @@
+
+        self.debug = debug
+        self.info = info
+        self.warning = warning
+        self.error = error
+
+
+class ImagesInterface:
+    def __init__(self, services: InvocationServices, context_data: InvocationContextData) -> None:
+        def save(
+            image: Image,
+            board_id: Optional[str] = None,
+            image_category: ImageCategory = ImageCategory.GENERAL,
+            metadata: Optional[MetadataField] = None,
+        ) -> ImageDTO:
+            """
+            Saves an image, returning its DTO.
+
+            If the current queue item has a workflow or metadata, it is automatically saved with the image.
+
+            :param image: The image to save, as a PIL image.
+            :param board_id: The board ID to add the image to, if it should be added.
+            :param image_category: The category of the image. Only the GENERAL category is added \
+                to the gallery.
+            :param metadata: The metadata to save with the image, if it should have any. If the \
+                invocation inherits from `WithMetadata`, that metadata will be used automatically. \
+                **Use this only if you want to override or provide metadata manually!**
+            """
+
+            # If the invocation inherits metadata, use that. Else, use the metadata passed in.
+            metadata_ = (
+                context_data.invocation.metadata if isinstance(context_data.invocation, WithMetadata) else metadata
+            )
+
+            return services.images.create(
+                image=image,
+                is_intermediate=context_data.invocation.is_intermediate,
+                image_category=image_category,
+                board_id=board_id,
+                metadata=metadata_,
+                image_origin=ResourceOrigin.INTERNAL,
+                workflow=context_data.workflow,
+                session_id=context_data.session_id,
+                node_id=context_data.invocation.id,
+            )
+
+        def get_pil(image_name: str) -> Image:
+            """
+            Gets an image as a PIL Image object.
+
+            :param image_name: The name of the image to get.
+            """
+            return services.images.get_pil_image(image_name)
+
+        def get_metadata(image_name: str) -> Optional[MetadataField]:
+            """
+            Gets an image's metadata, if it has any.
+
+            :param image_name: The name of the image to get the metadata for.
+            """
+            return services.images.get_metadata(image_name)
+
+        def get_dto(image_name: str) -> ImageDTO:
+            """
+            Gets an image as an ImageDTO object.
+
+            :param image_name: The name of the image to get.
+            """
+            return services.images.get_dto(image_name)
+
+        def update(
+            image_name: str,
+            board_id: Optional[str] = None,
+            is_intermediate: Optional[bool] = False,
+        ) -> ImageDTO:
```

Comment:
The method exists here because I was trying to dogfood the this new context _and_ still hide the full-fat services.

If we `A)` don't try to hide the services and/or `B)` add the `WithBoard` class and handling described below, this can be removed.

## Context

`LinearUIOutputInvocation` is used only by the linear UI, and is hidden from the nodes menu in the workflows UI. It adds images to boards and changes the `is_intermediate` status for already-saved images without needing to write the image a second time.

Why?

The VAE decode node doesn't have a `board_id` field, so its outputs are saved without a board association. To auto-save images to a board, we need to take some action after the VAE decode. We are on the third solution now.

### Three solutions

- First, we waited until an image was received, then hit the API to add that image to the board. This added some jank as images appeared first in the uncategorized area, then were moved to a board after a moment.
- Second, we added a `SaveImage` node that added the image to the board, but also re-saved the image. Graphs that used the node ended up with duplicate output images (only one of which was in the gallery). This was not popular.
- Third, `LinearUIOutputInvocation` is the same as `SaveImage`, but without writing the image a second time.

The intention with these solutions was to avoid adding a board field to every image-outputting node. I perceived this as tedious to maintain, and would add more complexity to the UI for an individual node.

### Proposal

At this point, though, I think it makes sense to add a `WithBoard` class that works like `WithMetadata`. When a node inherits from `WithMetadata`, it get the metadata field "for free", without needing to worry about handling the metadata unless it's needed.

Similarly, `WithBoard` would add a board dropdown to the node, and the context wrapper would handle grabbing the board selection and saving the image with it. The node author would get board-adding functionality by just inheriting, without any changes to their node logic.

I hadn't figured out this pattern at the time I did the other solutions.

Reply:
I've gone ahead and implemented `WithBoard` as described.

Reply:
There are a chunk of UI changes, which are only there to support this new way to save boards. Sorry it increased the PR size a bit.